### ISSUE TITLE
Payment fails when having surcharges with SwagCustomProducts

### ISSUE THREAD

**Issue by MelvinAchterhuis** [mollie/Shopware6#587]
When using [SwagCustomProducts](https://docs.shopware.com/en/shopware-6-en/extensions/customproducts) and having surcharges on a configuration the payment fails with the following error:

`Mollie.ERROR: Error when starting Mollie payment: Could not create Mollie order {"function":"order-prepare","exception":"[object] (RuntimeException(code: 422): Could not create Mollie order at /var/domains/domain.xyz/releases/20230623-114827-1a22d25984fd395fd43a13ab47f8d1ff9c4e508d/vendor/store.shopware.com/molliepayments/src/Service/MollieApi/Order.php:184)\n[previous exception] [object] (Mollie\\Api\\Exceptions\\ApiException(code: 422): [2023-06-23T12:06:38+0000] Error executing API call (422: Unprocessable Entity): Order line 1 is invalid. Total amount is off. Expected total amount to be €23.73 (1 × €23.73), got €27.86.`

This is what the lineItem looks like:

![image](https://github.com/mollie/Shopware6/assets/26538915/d3f26a67-1802-4bfa-96a1-e43c638ddd4b)

![image](https://github.com/mollie/Shopware6/assets/26538915/33f9d4a5-76f1-44b5-8bd2-f15c13f43042)

**Comment by HashtagAssist** [comment]
Same here, credit card payments failed. Even the new rounding error handling option dont work. We need this fixed asap because a lot of transactions failed and we lost customers. Thanks :)

**Comment by fjbender** [comment]
Hi all,

As far as I know we're blocked by this Shopware issue on this: https://issues.shopware.com/issues/CUS-539

Can you confirm this is still the case? The issue remains open, but it is pretty old.

Thanks

**Comment by HashtagAssist** [comment]
Not only with  one time surcharges. But also with custom products product options

**Comment by HashtagAssist** [comment]
I think i tracked it down.. Mabye its another issue but I think the vatRate of 1% from MelvinAchterhuis is a fake one to 

Example:

Total                                      5.451,52 €
Shipping costs                             0,00 €
Total without VAT                  5.031,59 €
plus 7% VAT                             312,70 €
plus 19% VAT                           107,23 €
Total sum incl. VAT                5.451,52 €

Total taxes:                                419.93€
Total expected by Mollie:          420.12€



The calculation probably takes place here
molliepayments/src/Service/MollieApi/PriceCalculator.php
in the function calculateLineItemPrice.

In line 89 a fictitious tax rate is calculated because it is a shopping cart with mixed tax rates. 
There, vatAmount is also calculated correctly (419.93). 
The simulated vatRate is 8.35.

If I calculate the tax amount from the total amount 5.451,52€ with 8,35%, I get the 420,12€ (wrong).

I suspect that in your Api backend I calculate the tax total independently based on the transferred vatRate and totalAmount. 
In the order.php, function createOrder there is an $orderData array with the property "lines" there is also the correct sum (419,93€) under "vatAmount".

**Comment by msestendrup** [comment]
Hi
I just wanted to let you know that we resolved the issue [CUS-539](https://issues.shopware.com/issues/CUS-539). If another issue occurs, please create a new ticket in the [issue tracker](https://issues.shopware.com).
Thank you.
